src/jank/clojure/core.jank

```diff
@@ -431,6 +431,92 @@
                  #{ o }#
                );"))
 
+(defn conj!
+  ([]
+   (transient []))
+  ([coll]
+   coll)
+  ([coll x]
+   (native/raw "__value = visit_object
+                (
+                  [=](auto const typed_t, auto const head) -> object_ptr
+                  {
+                    using T = typename decltype(typed_t)::value_type;
+
+                    if constexpr(behavior::persistentable<T>)
+                    { return typed_t->cons_in_place(head); }
+                    else
+                    { throw #{ (ex-info :not-persistentable {:o coll}) }#; }
+                   },
+                   #{ coll }#,
+                   #{ x }#
+                 );
+               ")))
+
+(defn assoc!
+  ([coll k v]
+   (native/raw "__value = visit_object
+                (
+                  [=](auto const typed_t, auto const key, auto const val) -> object_ptr
+                  {
+                    using T = typename decltype(typed_t)::value_type;
+
+                    if constexpr(behavior::associatively_writable_in_place<T>)
+                    { return typed_t->assoc_in_place(key, val); }
+                    else
+                    { throw #{ (ex-info :not-associatively_writable_in_place {:o coll}) }#; }
+                  },
+                  #{ coll }#,
+                  #{ k }#,
+                  #{ v }#
+                );
+               "))
+  ([coll k v & kvs]
+    (let [ret (assoc! coll k v)]
+      (if kvs
+        (recur ret (first kvs) (second kvs) (nnext kvs))
+        ret))))
+
+(defn dissoc!
+  ([coll k]
+   (native/raw "__value = visit_object
+                (
+                  [=](auto const typed_t, auto const key) -> object_ptr
+                  {
+                    using T = typename decltype(typed_t)::value_type;
+
+                    if constexpr(behavior::associatively_writable_in_place<T>)
+                    { return typed_t->dissoc_in_place(key); }
+                    else
+                    { throw #{ (ex-info :not-associatively_writable_in_place {:o coll}) }#; }
+                  },
+                  #{ coll }#,
+                  #{ k }#
+                );
+               "))
+  ([coll k & ks]
+    (let [ret (dissoc! coll k)]
+      (if ks
+        (recur ret (first ks) (next ks))
+        ret))))
+
+(defn pop!
+  ([coll]
+   (native/raw "__value = visit_object
+                (
+                  [=](auto const typed_t) -> object_ptr
+                  {
+                    using T = typename decltype(typed_t)::value_type;
+
+                    if constexpr(std::same_as<T, obj::transient_vector>)
+                    { return typed_t->pop_in_place(); }
+                    else
+                    { throw #{ (ex-info :not-transient-vector {:o coll}) }#; }
+                  },
+                  #{ coll }#
+                );
```

Comment:
`visit_object` is very expensive, at compile-time, since it results in an instantiation of the lambda for every object type (nearly 30 right now). For the other transient fns, we're supporting multiple object types using behaviors, so we're not checking for transient maps, sets, or vectors specifically. However, for `pop!`, we really only care about transient vectors. When we use `visit_object` for this, only one instantiation of the lambda is useful and all of the others just throw an exception. It would be much more efficient, at compile-time, to instead just check for `~{ coll }->type == object_type::transient_vector`.

The difference between a dozen `visit_object` calls versus explicit type checks can actually be multiple full seconds of compilation time. Since we JIT compile code in Clojure, compile-time performance can also mean run-time performance.

Tangentially, this is also why I have `visit_seqable`, which is like `visit_object`, but only instantiates the lambda for known seqable types. This is a very common thing we need to do and cutting down the number of those lambda instantiations to the known seqable types can significantly speed up compilation.